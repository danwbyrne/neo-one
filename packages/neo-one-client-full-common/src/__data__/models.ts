import { common, ContractParameterTypeModel, ECPoint, SignatureString, UInt160 } from '@neo-one/client-common';
import { constants } from '@neo-one/utils';
import {
  ContractABIModel,
  ContractEventModel,
  ContractFunctionModel,
  ContractGroupModel,
  ContractManifestModel,
  ContractModel,
  ContractParameterModel,
  ContractPermissionDescriptorModel,
  ContractPermissionsModel,
  ContractPropertyStateModel,
} from '../models';

export const contractParamModel = {
  boolean: new ContractParameterModel({
    type: ContractParameterTypeModel.Boolean,
    name: 'param',
  }),
  byteArray: new ContractParameterModel({
    type: ContractParameterTypeModel.ByteArray,
    name: 'param',
  }),
  hash160: new ContractParameterModel({
    type: ContractParameterTypeModel.Hash160,
    name: 'param',
  }),
  hash256: new ContractParameterModel({
    type: ContractParameterTypeModel.Hash256,
    name: 'param',
  }),
  array: new ContractParameterModel({
    type: ContractParameterTypeModel.Array,
    name: 'param',
  }),
  integer: new ContractParameterModel({
    type: ContractParameterTypeModel.Integer,
    name: 'param',
  }),
  interopInterface: new ContractParameterModel({
    type: ContractParameterTypeModel.InteropInterface,
    name: 'param',
  }),
  map: new ContractParameterModel({
    type: ContractParameterTypeModel.Map,
    name: 'param',
  }),
  publicKey: new ContractParameterModel({
    type: ContractParameterTypeModel.PublicKey,
    name: 'param',
  }),
  signature: new ContractParameterModel({
    type: ContractParameterTypeModel.Signature,
    name: 'param',
  }),
  string: new ContractParameterModel({
    type: ContractParameterTypeModel.String,
    name: 'param',
  }),
  void: new ContractParameterModel({
    type: ContractParameterTypeModel.Void,
    name: 'param',
  }),
};

export const contractFunctionModel = (
  parameters: readonly ContractParameterModel[] = [contractParamModel.boolean],
  returnType: ContractParameterTypeModel = ContractParameterTypeModel.Void,
  name = 'function',
) => new ContractFunctionModel({ name, parameters, returnType });

export const contractEventModel = (
  parameters: readonly ContractParameterModel[] = [contractParamModel.boolean],
  name = 'event',
) => new ContractEventModel({ name, parameters });

export const contractAbiModel = (
  methods: readonly ContractFunctionModel[] = [contractFunctionModel()],
  events: readonly ContractEventModel[] = [contractEventModel()],
  entryPoint: ContractFunctionModel = contractFunctionModel(),
  hash = common.bufferToUInt160(Buffer.alloc(20, 1)),
) => new ContractABIModel({ hash, entryPoint, methods, events });

export const contractGroupModel = (
  publicKey: ECPoint = common.stringToECPoint(constants.PRIVATE_NET_PUBLIC_KEY),
  signature: SignatureString = 'ccaab040cc25021c91567b75db4778853441869157b8f6aad960cdcf1069812480027a528ca9b98e2205027de20696f848cf81824eeb7af1d5110870870ceb67',
) => new ContractGroupModel({ publicKey, signature });

export const contractPermissionDescriptorModel = (hashOrGroupType: 'uint160' | 'ecpoint' | undefined) => {
  let hashOrGroup: UInt160 | ECPoint | undefined;
  if (hashOrGroupType === 'uint160') {
    hashOrGroup = common.bufferToUInt160(Buffer.alloc(20, 1));
  } else if (hashOrGroupType === 'ecpoint') {
    hashOrGroup = common.stringToECPoint(constants.PRIVATE_NET_PUBLIC_KEY);
  }

  return new ContractPermissionDescriptorModel({ hashOrGroup });
};

export const contractPermissionsModel = (
  hashOrGroupType: 'uint160' | 'ecpoint' | undefined,
  methods: readonly string[] = [],
) => new ContractPermissionsModel({ contract: contractPermissionDescriptorModel(hashOrGroupType), methods });

export const contractManifestModel = (
  groups: readonly ContractGroupModel[] = [contractGroupModel()],
  features: ContractPropertyStateModel = ContractPropertyStateModel.HasStoragePayable,
  abi: ContractABIModel = contractAbiModel(),
  permissions: readonly ContractPermissionsModel[] = [contractPermissionsModel('uint160', ['method1'])],
  trusts: readonly UInt160[] = [common.bufferToUInt160(Buffer.alloc(20, 1))],
  safeMethods: readonly string[] = ['method1', 'method2'],
) => new ContractManifestModel({ groups, features, abi, permissions, trusts, safeMethods });

export const contractModel = (
  script: Buffer = Buffer.alloc(25),
  manifest: ContractManifestModel = contractManifestModel(),
) => new ContractModel({ script, manifest });
