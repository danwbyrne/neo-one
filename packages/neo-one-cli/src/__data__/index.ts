export * from './getClients';
export * from './getContracts';
export * from './buildTests';
