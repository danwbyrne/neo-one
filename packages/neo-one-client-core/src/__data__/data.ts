import BN from 'bn.js';

const buffers = {
  a: 'b500',
  b: '5f8d70',
};
const hash256s = {
  a: '0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234',
  b: '0x0621113df436c180a47b833a2814008dc4f332915b22253651593a40a342e0fb',
  c: '0x6378f79eadb05aa5ef5bca489330e059305a1180bd69021a749f63c56e37c834',
  d: '0x5f8d7c7ec5c984cb632f4d8a5c867271197dee4a09f37cbab95d799cb292cbd1',
  e: '0x669df446205866859074e6f5658479f9f25616bc682c2e5b8e28b7dfbcf6c288',
};
const bns = {
  a: new BN(1_00000000),
  b: new BN(10_00000000),
};
const signatures = {
  a:
    'b50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234b50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234',
};

export const data = {
  buffers,
  bns,
  hash256s,
  signatures,
};
