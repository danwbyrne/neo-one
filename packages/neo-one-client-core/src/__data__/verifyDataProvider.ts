import {
  ActionJSON,
  AddressString,
  Asset,
  AssetJSON,
  AssetNameJSON,
  AssetType,
  AssetTypeJSON,
  CallReceiptJSON,
  ConfirmedTransaction,
  Contract,
  ContractJSON,
  InvocationResultJSON,
  Output,
  OutputJSON,
  PublicKeyString,
  RawAction,
  RawCallReceipt,
  RawInvocationResult,
  RawInvocationResultJSON,
  scriptHashToAddress,
  Transaction,
  TransactionJSON,
  VMState,
} from '@neo-one/client-common';
import BigNumber from 'bignumber.js';

interface AssetBase {
  readonly type: AssetType;
  readonly name: string;
  readonly amount: BigNumber;
  readonly precision: number;
  readonly owner: PublicKeyString;
  readonly admin: AddressString;
}

interface AssetBaseJSON {
  readonly type: AssetTypeJSON;
  readonly name: AssetNameJSON;
  readonly amount: string;
  readonly precision: number;
  readonly owner: string;
  readonly admin: string;
}

const verifyInvocationResultSuccess = (
  invocationResult: RawInvocationResult,
  invocationResultJSON: InvocationResultJSON | RawInvocationResultJSON,
) => {
  if (
    invocationResult.state !== 'HALT' ||
    (invocationResultJSON.state !== VMState.HALT && invocationResultJSON.state !== 'HALT')
  ) {
    throw new Error('For TS');
  }
  expect(invocationResult.gasConsumed.toString(10)).toEqual(invocationResultJSON.gas_consumed);
  expect(invocationResult.gasCost.toString(10)).toEqual(invocationResultJSON.gas_cost);
  const firstStack = invocationResult.stack[0];
  const firstStackJSON = invocationResultJSON.stack[0];
  if (firstStack.type !== 'Integer' || firstStackJSON.type !== 'Integer') {
    throw new Error('For TS');
  }
  expect(firstStack.value.toString(10)).toEqual(firstStackJSON.value);
};

const verifyDefaultActions = (
  actions: readonly RawAction[],
  actionsJSON: readonly ActionJSON[],
  blockIndex: number,
  blockHash: string,
  index: number,
  txid: string,
) => {
  expect(actions.length).toEqual(actionsJSON.length);
  const verifyAction = (actionResult: RawAction, action: ActionJSON, idx: number) => {
    expect(actionResult.type).toEqual(action.type);
    expect(actionResult.version).toEqual(action.version);
    expect(actionResult.blockIndex).toEqual(blockIndex);
    expect(actionResult.blockHash).toEqual(blockHash);
    expect(actionResult.transactionIndex).toEqual(index);
    expect(actionResult.transactionHash).toEqual(txid);
    expect(actionResult.index).toEqual(idx);
    expect(actionResult.globalIndex.toString(10)).toEqual(action.index);
    expect(actionResult.address).toEqual(scriptHashToAddress(action.scriptHash));
  };
  verifyAction(actions[0], actionsJSON[0], 0);
  verifyAction(actions[1], actionsJSON[1], 1);
};

const verifyAssetBase = (
  asset: AssetBase,
  assetJSON: AssetBaseJSON,
  toAssetType: string = assetJSON.type,
  name = assetJSON.name,
) => {
  expect(asset.type).toEqual(toAssetType);
  expect(asset.name).toEqual(name);
  expect(asset.amount.toString(10)).toEqual(assetJSON.amount);
  expect(asset.precision).toEqual(assetJSON.precision);
  expect(asset.owner).toEqual(assetJSON.owner);
  expect(asset.admin).toEqual(assetJSON.admin);
};

const verifyAsset = (
  asset: Asset,
  assetJSON: AssetJSON,
  toAssetType: string = assetJSON.type,
  name = assetJSON.name,
) => {
  verifyAssetBase(asset, assetJSON, toAssetType, name);
  expect(asset.hash).toEqual(assetJSON.id);
  expect(asset.available.toString(10)).toEqual(assetJSON.available);
  expect(asset.issuer).toEqual(assetJSON.issuer);
  expect(asset.expiration).toEqual(assetJSON.expiration);
  expect(asset.frozen).toEqual(assetJSON.frozen);
};

const verifyContract = (contract: Contract, contractJSON: ContractJSON, returnType = 'Buffer') => {
  expect(contract.version).toEqual(contractJSON.version);
  expect(contract.address).toEqual(scriptHashToAddress(contractJSON.hash));
  expect(contract.script).toEqual(contractJSON.script);
  expect(contract.parameters).toEqual(['Address', 'Buffer']);
  expect(contract.returnType).toEqual(returnType);
  expect(contract.name).toEqual(contractJSON.name);
  expect(contract.codeVersion).toEqual(contractJSON.code_version);
  expect(contract.author).toEqual(contractJSON.author);
  expect(contract.email).toEqual(contractJSON.email);
  expect(contract.description).toEqual(contractJSON.description);
  expect(contract.storage).toEqual(contractJSON.properties.storage);
  expect(contract.dynamicInvoke).toEqual(contractJSON.properties.dynamic_invoke);
  expect(contract.payable).toEqual(contractJSON.properties.payable);
};

const verifyOutput = (output: Output, outputJSON: OutputJSON) => {
  expect(output.asset).toEqual(outputJSON.asset);
  expect(output.value.toString(10)).toEqual(outputJSON.value);
  expect(output.address).toEqual(outputJSON.address);
};

const verifyTransactionBase = (transaction: Transaction, transactionJSON: TransactionJSON) => {
  expect(transaction.hash).toEqual(transactionJSON.txid);
  expect(transaction.size).toEqual(transactionJSON.size);
  expect(transaction.version).toEqual(transactionJSON.version);
  expect(transaction.attributes.length).toEqual(transactionJSON.attributes.length);
  expect(transaction.attributes[0].usage).toEqual(transactionJSON.attributes[0].usage);
  expect(transaction.attributes[0].data).toEqual(scriptHashToAddress(transactionJSON.attributes[0].data));
  expect(transaction.attributes[1].usage).toEqual(transactionJSON.attributes[1].usage);
  expect(transaction.attributes[1].data).toEqual(transactionJSON.attributes[1].data);
  expect(transaction.attributes[2].usage).toEqual(transactionJSON.attributes[2].usage);
  expect(transaction.attributes[2].data).toEqual(transactionJSON.attributes[2].data);
  expect(transaction.attributes[3].usage).toEqual(transactionJSON.attributes[3].usage);
  expect(transaction.attributes[3].data).toEqual(transactionJSON.attributes[3].data);
  expect(transaction.inputs.length).toEqual(transactionJSON.vin.length);
  expect(transaction.inputs[0].hash).toEqual(transactionJSON.vin[0].txid);
  expect(transaction.inputs[0].index).toEqual(transactionJSON.vin[0].vout);
  expect(transaction.outputs.length).toEqual(transactionJSON.vout.length);
  verifyOutput(transaction.outputs[0], transactionJSON.vout[0]);
  expect(transaction.scripts).toEqual(transactionJSON.scripts);
  expect(transaction.systemFee.toString(10)).toEqual(transactionJSON.sys_fee);
  expect(transaction.networkFee.toString(10)).toEqual(transactionJSON.net_fee);
};

const verifyClaimTransaction = (transaction: Transaction, transactionJSON: TransactionJSON) => {
  verifyTransactionBase(transaction, transactionJSON);
  if (transaction.type !== 'ClaimTransaction' || transactionJSON.type !== 'ClaimTransaction') {
    throw new Error('For TS');
  }
  expect(transaction.claims.length).toEqual(transactionJSON.claims.length);
  expect(transaction.claims[0].hash).toEqual(transactionJSON.claims[0].txid);
  expect(transaction.claims[0].index).toEqual(transactionJSON.claims[0].vout);
};

const verifyContractTransaction = (transaction: Transaction, transactionJSON: TransactionJSON) => {
  verifyTransactionBase(transaction, transactionJSON);
  expect(transaction.type).toEqual('ContractTransaction');
};

const verifyEnrollmentTransaction = (transaction: Transaction, transactionJSON: TransactionJSON) => {
  verifyTransactionBase(transaction, transactionJSON);
  if (transaction.type !== 'EnrollmentTransaction' || transactionJSON.type !== 'EnrollmentTransaction') {
    throw new Error('For TS');
  }
  expect(transaction.publicKey).toEqual(transactionJSON.pubkey);
};

const verifyInvocationTransaction = (transaction: Transaction, transactionJSON: TransactionJSON) => {
  verifyTransactionBase(transaction, transactionJSON);
  if (transaction.type !== 'InvocationTransaction' || transactionJSON.type !== 'InvocationTransaction') {
    throw new Error('For TS');
  }
  expect(transaction.script).toEqual(transactionJSON.script);
  expect(transaction.gas.toString(10)).toEqual(transactionJSON.gas);
};

const verifyIssueTransaction = (transaction: Transaction, transactionJSON: TransactionJSON) => {
  verifyTransactionBase(transaction, transactionJSON);
  expect(transaction.type).toEqual('IssueTransaction');
};

const verifyMinerTransaction = (transaction: Transaction, transactionJSON: TransactionJSON) => {
  verifyTransactionBase(transaction, transactionJSON);
  if (transaction.type !== 'MinerTransaction' || transactionJSON.type !== 'MinerTransaction') {
    throw new Error('For TS');
  }
  expect(transaction.nonce).toEqual(transactionJSON.nonce);
};

const verifyPublishTransaction = (transaction: Transaction, transactionJSON: TransactionJSON) => {
  verifyTransactionBase(transaction, transactionJSON);
  if (transaction.type !== 'PublishTransaction' || transactionJSON.type !== 'PublishTransaction') {
    throw new Error('For TS');
  }
  verifyContract(transaction.contract, transactionJSON.contract);
};

const verifyRegisterTransaction = (transaction: Transaction, transactionJSON: TransactionJSON) => {
  verifyTransactionBase(transaction, transactionJSON);
  if (transaction.type !== 'RegisterTransaction' || transactionJSON.type !== 'RegisterTransaction') {
    throw new Error('For TS');
  }
  verifyAssetBase(transaction.asset, transactionJSON.asset);
};

const verifyStateTransaction = (transaction: Transaction, transactionJSON: TransactionJSON) => {
  verifyTransactionBase(transaction, transactionJSON);
  expect(transaction.type).toEqual('StateTransaction');
};

const verifyConfirmedTransaction = (transaction: ConfirmedTransaction, transactionJSON: TransactionJSON) => {
  if (transactionJSON.data === undefined) {
    throw new Error('For TS');
  }

  expect(transaction.receipt.blockHash).toEqual(transactionJSON.data.blockHash);
  expect(transaction.receipt.blockIndex).toEqual(transactionJSON.data.blockIndex);
  expect(transaction.receipt.transactionIndex).toEqual(transactionJSON.data.transactionIndex);
  expect(transaction.receipt.globalIndex.toString(10)).toEqual(transactionJSON.data.globalIndex);
};

const verifyCallReceipt = (receipt: RawCallReceipt, receiptJSON: CallReceiptJSON) => {
  verifyInvocationResultSuccess(receipt.result, receiptJSON.result);
  verifyDefaultActions(
    receipt.actions,
    receiptJSON.actions,
    0,
    '0x​​​​​0000000000000000000000000000000000000000000000000000000000000000​​​​​',
    0,
    '0x​​​​​0000000000000000000000000000000000000000000000000000000000000000​​​​​',
  );
};

export const verifyDataProvider = {
  verifyInvocationResultSuccess,
  verifyDefaultActions,
  verifyAssetBase,
  verifyAsset,
  verifyContract,
  verifyOutput,
  verifyTransactionBase,
  verifyClaimTransaction,
  verifyContractTransaction,
  verifyEnrollmentTransaction,
  verifyInvocationTransaction,
  verifyIssueTransaction,
  verifyMinerTransaction,
  verifyPublishTransaction,
  verifyRegisterTransaction,
  verifyStateTransaction,
  verifyConfirmedTransaction,
  verifyCallReceipt,
};
