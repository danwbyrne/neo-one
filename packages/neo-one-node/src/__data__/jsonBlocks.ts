export const genesisJSON = {
  hash: '0xc359030132be10fd19cfd0a27e289fe04acb0c5c4ca5254af8a2d99498c7da45',
  size: 172,
  version: 0,
  previousblockhash: '0x0000000000000000000000000000000000000000000000000000000000000000',
  merkleroot: '0x948aad0908a742eec79c52a0a4d5f95b1683a9adc3c2700211f425d49693878d',
  time: 1468595301000,
  index: 0,
  nextconsensus: 'NgPkjjLTNcQad99iRYeXRUuowE4gxLAnDL',
  witnesses: [
    {
      invocation: '',
      verification: 'EQ==',
    },
  ],
  consensusdata: {
    primary: 0,
    nonce: '000000007c2bac1d',
  },
  tx: [
    {
      hash: '0x173dcbc4a88995a0cf7bdd006923d148f787f76ca75621dc4c440ca6d9afbc73',
      size: 58,
      version: 0,
      nonce: 0,
      sender: 'NeN4xPMn4kHoj7G8Lciq9oorgLTvqt4qi1',
      sysfee: '0',
      netfee: '0',
      validuntilblock: 0,
      signers: [
        {
          account: '0xa7213b15cc18d19c810f644e37411d882ee561ca',
          scopes: 'FeeOnly', // FeeOnly should be None. This indicates that this JSON data is from Preview 2 or earlier, not Preview 3
        },
      ],
      attributes: [],
      script: 'QRI+f+g=',
      witnesses: [
        {
          invocation: '',
          verification: 'EQ==',
        },
      ],
    },
  ],
  confirmations: 291297,
  nextblockhash: '0x8f30c34a8a8ef997155aa4a0ef6664d872a127ecc3bb6a85b7bbc55d6d9912f5',
};

export const secondBlockJSON = {
  hash: '0x8f30c34a8a8ef997155aa4a0ef6664d872a127ecc3bb6a85b7bbc55d6d9912f5',
  size: 700,
  version: 0,
  previousblockhash: '0xc359030132be10fd19cfd0a27e289fe04acb0c5c4ca5254af8a2d99498c7da45',
  merkleroot: '0x9ff5ba403c81151c0d031b548fe6cff82ef61a01be72c38ecc81d4bc2f1ee01c',
  time: 1596537327793,
  index: 1,
  nextconsensus: 'NgPkjjLTNcQad99iRYeXRUuowE4gxLAnDL',
  witnesses: [
    {
      invocation:
        'DED/2tpnw4uN5407xKQuwAXw+Hm9L4P51hSMfjwbEcm7pN+aFdn5+d/VxT9ifDX0KQRGlbericqr6h2gQvnJYvHqDECO9w/7PAccs3K3yqAZ4zDTpsnxxDU5EyR1PNNcBUx31lFGyHtjcxhce+YNYxcJc2zOFFk+lFmCBn+ZiF1HZrkPDEBbpwRGODeexXqSws2VQPrOg/BiCNYKmAj3vPW8HvC3ypnly/Hfy8lYF9iZiIighzUALC4QzQpxdq1IdNMSH22jDEAe/l0FF8HXc6e20zT6x0D8XvrITHsKMxxe4S5hGYwojgBDjJPcMwvJbT83AF+sorbZc7SMaCGwT1XOEg+9M0G2DEDOK/OsvMqyKgyHf/gz5QFgQxPjHKuuMoN38+OLuMRP36K4Eunv7oql3KstrgpVIqZHtETEHmxMEFIFXu5hFEa8',
      verification:
        'FQwhAwCbdUDhDyVi5f2PrJ6uwlFmpYsm5BI0j/WoaSe/rCKiDCEDAgXpzvrqWh38WAryDI1aokaLsBSPGl5GBfxiLIDmBLoMIQIUuvDO6jpm8X5+HoOeol/YvtbNgua7bmglAYkGX0T/AQwhAj6bMuqJuU0GbmSbEk/VDjlu6RNp6OKmrhsRwXDQIiVtDCEDQI3NQWOW9keDrFh+oeFZPFfZ/qiAyKahkg6SollHeAYMIQKng0vpsy4pgdFXy1u9OstCz9EepcOxAiTXpE6YxZEPGwwhAroscPWZbzV6QxmHBYWfriz+oT4RcpYoAHcrPViKnUq9FwtBE43vrw==',
    },
  ],
  consensusdata: {
    primary: 1,
    nonce: 'f490dfe92971e572',
  },
  tx: [],
  confirmations: 294767,
  nextblockhash: '0x84fa4b26af4b00095d66714a3a2ec63acdeff7ae27ca5d4f6cbe5db2691d7fae',
};

export const thirdBlockJSON = {
  hash: '0x84fa4b26af4b00095d66714a3a2ec63acdeff7ae27ca5d4f6cbe5db2691d7fae',
  size: 700,
  version: 0,
  previousblockhash: '0x8f30c34a8a8ef997155aa4a0ef6664d872a127ecc3bb6a85b7bbc55d6d9912f5',
  merkleroot: '0x905819ef1743d2f7e522dacc90a365dd7e872ebdee74f2dfcad61d5089dc9f64',
  time: 1596537468402,
  index: 2,
  nextconsensus: 'NgPkjjLTNcQad99iRYeXRUuowE4gxLAnDL',
  witnesses: [
    {
      invocation:
        'DEC3IxKTiG/yHShrtaN2aDBWLCoiLCvaUHmir1TovNGookWVKqB6NxwJb1capeJk50HtWntpz4B/k0oDd9Y33F8xDEDikpijBQ+rl4WJCHN7DPumv8OvtBhszmDKqX3wEJyaKAUL8rLsncDbsgIwJTxPyMUDIrU8BbVIEYDAH0KvB5AfDECcu9VKxSaC5mJdWvwguqw74ZU31UIG2cn5FMGg/EPMwFQeJEW/HTZNIIJNvE8lQ5i6C7wZ5QJFRLXavqDwvO+TDEBokh+09rXIVbcmzGBqddJ+Kq9mp7UAH9XTcTtDiQaeKQ39krUyySuaUjZm54kHsi70OsAqkK+hfSIpYZ4qKLeUDEA5qVvZMvHzv9NR6ekPvKpTquy32qgY+quCJt1BSeBaOGAZ4wTVs3Guw+HvFfarRPORYy186jVzQM1tAbPR+oiQ',
      verification:
        'FQwhAwCbdUDhDyVi5f2PrJ6uwlFmpYsm5BI0j/WoaSe/rCKiDCEDAgXpzvrqWh38WAryDI1aokaLsBSPGl5GBfxiLIDmBLoMIQIUuvDO6jpm8X5+HoOeol/YvtbNgua7bmglAYkGX0T/AQwhAj6bMuqJuU0GbmSbEk/VDjlu6RNp6OKmrhsRwXDQIiVtDCEDQI3NQWOW9keDrFh+oeFZPFfZ/qiAyKahkg6SollHeAYMIQKng0vpsy4pgdFXy1u9OstCz9EepcOxAiTXpE6YxZEPGwwhAroscPWZbzV6QxmHBYWfriz+oT4RcpYoAHcrPViKnUq9FwtBE43vrw==',
    },
  ],
  consensusdata: {
    primary: 1,
    nonce: '3b23f5541d704aaa',
  },
  tx: [],
  confirmations: 311250,
  nextblockhash: '0xfd4d6fdf33eaec2d574590af1c8cbfcaa9d990b4ecc1cf0e60670b951be214a2',
};
