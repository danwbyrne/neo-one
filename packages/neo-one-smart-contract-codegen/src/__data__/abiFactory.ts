import { ABI, ABIDefault, ABIReturn } from '@neo-one/client-common';

// tslint:disable-next-line:no-any
const getCombinations = (list: readonly (readonly any[])[]) => {
  const floorsIn = list.reduceRight<readonly number[]>((acc, right) => [right.length * acc[0]].concat(acc), [1]);
  const max = floorsIn[0];
  const floors = floorsIn.slice(1);

  const iteratorFuncs = list.map((item, index) =>
    index === list.length - 1
      ? (value: number) => item[value % item.length]
      : (value: number) => item[Math.floor(value / floors[index]) % item.length],
  );

  return [...Array(max)].map((_, index) => iteratorFuncs.map((func) => func(index)));
};

const abiFunctionOptions: readonly { readonly [key in string]?: boolean }[] = [
  { constant: true },
  { send: true },
  { sendUnsafe: true },
  { sendUnsafe: true, receive: true },
  { receive: true },
  { claim: true },
  { refundAssets: true },
  { completeSend: true },
  {},
];

const abiParameterOptions: readonly { readonly [key in string]?: boolean | ABIDefault }[] = [
  { default: { type: 'sender' } },
  { rest: true },
  { default: { type: 'sender' }, rest: true },
  {},
];

const abiReturns: readonly ABIReturn[] = [
  {
    type: 'ForwardValue',
  },
  {
    type: 'Boolean',
    optional: true,
  },
  {
    type: 'Boolean',
    forwardedValue: true,
  },
  {
    type: 'Integer',
    decimals: 4,
  },
  {
    type: 'Array',
    value: {
      type: 'Boolean',
    },
  },
  {
    type: 'Map',
    key: { type: 'Integer', decimals: 0 },
    value: { type: 'Boolean' },
  },
  {
    type: 'Object',
    properties: {
      property: { type: 'Boolean' },
    },
  },
];

const boolReturnType: ABIReturn = { type: 'Boolean' };

const createBaseABIFunctions = (): readonly ABI[] => {
  const combinations = getCombinations([abiReturns, abiFunctionOptions]);

  return combinations.map(([abiReturn, abiFuncProps]) => {
    const funcPropNames = Object.keys(abiFuncProps);
    const name = `${abiReturn.type}${`${funcPropNames.length === 0 ? '' : '_'}${funcPropNames.join('_')}`}`;

    return {
      functions: [
        {
          name,
          returnType: abiReturn,
          ...abiFuncProps,
        },
      ],
    };
  });
};

const createBaseABIEvents = (): readonly ABI[] => {
  const combinations = getCombinations([abiReturns, abiParameterOptions]);

  return combinations.map(([abiReturn, abiParamProps]) => {
    const paramPropNames = Object.keys(abiParamProps);
    const name = `${abiReturn.type}${`${paramPropNames.length === 0 ? '' : '_'}${paramPropNames.join('_')}`}`;

    return {
      functions: [
        {
          name: 'foo',
          returnType: boolReturnType,
        },
      ],
      events: [
        {
          name,
          parameters: [
            {
              name: 'arg',
              ...abiReturn,
              ...abiParamProps,
            },
          ],
        },
      ],
    };
  });
};

const createBaseABIFunctionParameters = (): readonly ABI[] => {
  const combinations = getCombinations([abiReturns, abiParameterOptions]);

  return combinations.map(([abiReturn, abiParamProps]) => {
    const paramPropNames = Object.keys(abiParamProps);
    const name = `${abiReturn.type}${`${paramPropNames.length === 0 ? '' : '_'}${paramPropNames.join('_')}`}`;

    return {
      functions: [
        {
          name,
          parameters: [
            {
              name: 'arg',
              ...abiReturn,
              ...abiParamProps,
            },
          ],
          returnType: boolReturnType,
        },
      ],
    };
  });
};

export const abiFactory = {
  createBaseABIFunctions,
  createBaseABIFunctionParameters,
  createBaseABIEvents,
};
