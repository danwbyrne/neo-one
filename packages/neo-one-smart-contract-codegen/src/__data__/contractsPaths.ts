import { ContractPaths } from '../type';

const sourceMap = {
  version: 0,
  sources: [],
  names: [],
  mappings: 'sourcemap',
  file: 'file',
};

export const contractsPaths: readonly ContractPaths[] = [
  {
    name: 'Token',
    contractPath: '/foo/bar/one/contracts/Token.ts',
    typesPath: '/foo/bar/one/generated/Token/types.js',
    createContractPath: '/foo/bar/one/generated/Token/contract.js',
    abiPath: '/foo/bar/one/generated/Token/abi.js',
    addresses: ['AV6PhAq1FHBeCMsNUPMWzpUFxvXRFwHoLn'],
    sourceMap,
  },
  {
    name: 'ICO',
    contractPath: '/foo/bar/one/contracts/ICO.ts',
    typesPath: '/foo/bar/one/generated/ICO/types.js',
    createContractPath: '/foo/bar/one/generated/ICO/contract.js',
    abiPath: '/foo/bar/one/generated/ICO/abi.js',
    addresses: ['ATfAohZQ5iuGDXWjcHxLAj8WXwQchGLXhi'],
    sourceMap,
  },
];
