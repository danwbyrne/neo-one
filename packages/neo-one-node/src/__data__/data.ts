import { BlockJSON, common, crypto, JSONHelper, toWitnessScope, WitnessScopeModel } from '@neo-one/client-common';
import { Block, ConsensusData, Signer, Transaction, Witness } from '@neo-one/node-core';
import { BN } from 'bn.js';
import { genesisJSON, secondBlockJSON, thirdBlockJSON } from './jsonBlocks';

const convertBlock = (json: BlockJSON) =>
  new Block({
    version: json.version,
    previousHash: JSONHelper.readUInt256(json.previousblockhash),
    timestamp: new BN(json.time),
    index: json.index,
    nextConsensus: crypto.addressToScriptHash({
      addressVersion: common.NEO_ADDRESS_VERSION,
      address: json.nextconsensus,
    }),
    merkleRoot: JSONHelper.readUInt256(json.merkleroot),
    witness: new Witness({
      invocation: Buffer.from(json.witnesses[0].invocation, 'base64'),
      verification: Buffer.from(json.witnesses[0].verification, 'base64'),
    }),
    transactions: json.tx.map(
      (tx) =>
        new Transaction({
          hash: JSONHelper.readUInt256(tx.hash),
          nonce: tx.nonce,
          attributes: [],
          script: Buffer.from(tx.script, 'base64'),
          systemFee: new BN(tx.sysfee),
          networkFee: new BN(tx.netfee),
          validUntilBlock: tx.validuntilblock,
          witnesses: tx.witnesses.map(
            (witness) =>
              new Witness({
                invocation: Buffer.from(witness.invocation, 'base64'),
                verification: Buffer.from(witness.verification, 'base64'),
              }),
          ),
          version: tx.version,
          signers: tx.signers.map(
            (signer) =>
              new Signer({
                account: JSONHelper.readUInt160(signer.account),
                scopes: (signer.scopes as any) === 'FeeOnly' ? WitnessScopeModel.None : toWitnessScope(signer.scopes),
              }),
          ),
        }),
    ),
    consensusData: json.consensusdata
      ? new ConsensusData({
          primaryIndex: json.consensusdata.primary,
          nonce: new BN(json.consensusdata.nonce, 16),
        })
      : undefined,
  });

const genesisBlockFromJSON = convertBlock(genesisJSON as any);
const secondBlockFromJSON = convertBlock(secondBlockJSON as any);
const thirdBlockFromJSON = convertBlock(thirdBlockJSON as any);

export const data = {
  genesisBlock: genesisBlockFromJSON,
  secondBlock: secondBlockFromJSON,
  thirdBlock: thirdBlockFromJSON,
};
