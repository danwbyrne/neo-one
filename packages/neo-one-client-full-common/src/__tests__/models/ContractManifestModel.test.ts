import { BinaryWriter } from '@neo-one/client-common';
import { contractManifestModel } from '../../__data__';

describe('ContractManifestModel - serializeWireBase', () => {
  let writer: BinaryWriter;
  beforeEach(() => {
    writer = new BinaryWriter();
  });

  test('Simple', () => {
    contractManifestModel().serializeWireBase(writer);
    expect(writer.toBuffer()).toMatchSnapshot();
  });
});
