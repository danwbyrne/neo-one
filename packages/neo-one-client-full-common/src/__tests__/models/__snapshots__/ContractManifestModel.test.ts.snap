// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractManifestModel - serializeWireBase Simple 1`] = `"01010101010101010101010101010101010101010866756e6374696f6e010105706172616d01ff0866756e6374696f6e010105706172616d01ff056576656e74010105706172616d0248be3c070df745a60f3b8b494efcc6caf90244d803a9a72fe95d9bae2120ec708063636161623034306363323530323163393135363762373564623437373838353334343138363931353762386636616164393630636463663130363938313234383030323761353238636139623938653232303530323764653230363936663834386366383138323465656237616631643531313038373038373063656236370101010101010101010101010101010101010101076d6574686f64310101010101010101010101010101010101010101076d6574686f6431076d6574686f643205"`;
